projects/RabbitMQ.Client/client/impl/OutgoingCommand.cs

```diff
@@ -31,49 +31,75 @@
 
 using System;
 using System.Buffers;
-using RabbitMQ.Client.Framing.Impl;
 
 namespace RabbitMQ.Client.Impl
 {
-    internal readonly struct OutgoingCommand
+    internal interface IOutgoingCommand
     {
-        // EmptyFrameSize, 8 = 1 + 2 + 4 + 1
-        // - 1 byte of frame type
-        // - 2 bytes of channel number
-        // - 4 bytes of frame payload length
-        // - 1 byte of payload trailer FrameEnd byte
-        private const int EmptyFrameSize = 8;
+        MethodBase Method { get; }
+        ReadOnlyMemory<byte> Transmit(ushort channelNumber, uint frameMax);
+    }
 
-        public readonly MethodBase Method;
-        private readonly ContentHeaderBase _header;
-        private readonly ReadOnlyMemory<byte> _body;
+    internal readonly struct OutgoingCommand : IOutgoingCommand
+    {
+        private readonly MethodBase _method;
+        public MethodBase Method => _method;
 
         public OutgoingCommand(MethodBase method)
-            : this(method, null, ReadOnlyMemory<byte>.Empty)
         {
+            _method = method;
         }
 
-        public OutgoingCommand(MethodBase method, ContentHeaderBase header, ReadOnlyMemory<byte> body)
+        public ReadOnlyMemory<byte> Transmit(ushort channelNumber, uint frameMax)
         {
-            Method = method;
+            int size = Method.GetRequiredBufferSize() + Framing.Method.FrameSize;
+
+            // Will be returned by SocketFrameWriter.WriteLoop
+            byte[] rentedArray = ArrayPool<byte>.Shared.Rent(size);
+            Span<byte> span = rentedArray.AsSpan(0, size);
+            Framing.Method.WriteTo(span, channelNumber, Method);
+#if DEBUG
+            if (offset != size)
+            {
+                throw new InvalidOperationException($"Serialized to wrong size, expect {size}, offset {offset}");
+            }
+#endif
+
+            return new ReadOnlyMemory<byte>(rentedArray, 0, size);
+        }
+    }
+
+    internal readonly struct OutgoingContentCommand : IOutgoingCommand
+    {
+        public MethodBase Method => _method;
+        private readonly MethodBase _method;
+        private readonly ContentHeaderBase _header;
+        private readonly ReadOnlyMemory<byte> _body;
+
+        public OutgoingContentCommand(MethodBase method, ContentHeaderBase header, ReadOnlyMemory<byte> body)
+        {
+            _method = method;
             _header = header;
             _body = body;
         }
 
-        internal void Transmit(ushort channelNumber, Connection connection)
+        public ReadOnlyMemory<byte> Transmit(ushort channelNumber, uint frameMax)
```

Comment:
Is transmit still a good name since we don't really transmit but more Prepare or create the payload

Reply:
Yeah, something like `SerializeToFrames` would probably be more descriptive.